### PR TITLE
Allows instanceOf and typeOf matchers to use the casted value

### LINKED ISSUE #666: isInstanceOf matcher should allow lamba for further tests

### PR DESCRIPTION
**Pull request by LeoColman**
This commit also adds some documentation to all related methods.

Fix #666

### CODE REVIEW

**Comment by LeoColman** [comment]
@dave08

File: kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/types/TypeMatchersTest.kt [hunk 1]
```diff
@@ -66,6 +56,14 @@ class TypeMatchersTest : WordSpec() {
           arrayList.shouldNotBeInstanceOf<ArrayList<*>>()
         }
       }
+      
+      "Allow execution with a lambda" {
+        val list = arrayListOf(1, 2, 3)
+        
+        list.shouldBeInstanceOf<ArrayList<Int>> { it: ArrayList<Int> ->
```
Comment (LeoColman) [inline kotlintest-tests/kotlintest-tests-core/src/test/kotlin/com/sksamuel/kotlintest/matchers/types/TypeMatchersTest.kt:63]: type was made explicit to enforce that this type won't change in the future, even though it's redundant for users

**Comment by sksamuel** [comment]
This allows us to do `t.shouldBeInstanceOf<T> { t -> ... }` ?

**Comment by LeoColman** [comment]
Yes. An example of this is on the test. I made the type explicit to ensure it won't change in the future

**Review by sksamuel — APPROVED** [review]
Good work as always.

**Pull request merged**